### PR TITLE
Add support for WHERE LIKE ANY queries

### PR DESCRIPTION
**Pull request by 3ch01c**
See #3332.

### CODE REVIEW

File: test/integration/dialects/postgres/query-generator.test.js [hunk 1]
```diff
@@ -394,6 +394,10 @@ if (dialect.match(/^postgres/)) {
           arguments: ['myTable', {where: { field: new Buffer('Sequelize')}}],
           expectation: "SELECT * FROM \"myTable\" WHERE \"myTable\".\"field\" = E'\\\\x53657175656c697a65';",
           context: QueryGenerator
+        }, {
+          title: 'use LIKE ANY when using LIKE operator and value has "any" key',
+          arguments: ['myTable', {where: {field: {like: {any: ['foo', 'bar', 'baz']}}}}],
+          expectation: "SELECT * FROM \"myTable\" WHERE \"myTable\".\"field\" LIKE ANY (ARRAY['foo','bar','baz']);",
```
Comment (mickhansen) [inline test/integration/dialects/postgres/query-generator.test.js]: Please add tests to `sql/unit/sql/where` instead.

**Comment by 3ch01c** [comment]
@mickhansen Fail seems to be unrelated to this commit. Could you re-run Travis?

File: lib/dialects/abstract/query-generator.js [hunk 2]
```diff
@@ -1766,6 +1767,11 @@ module.exports = (function() {
         value = (value.$between || value.$notBetween).map(function (item) {
           return self.escape(item);
         }).join(' AND ');
+      } else if (value && ((value.$like && (any = value.$like.any) && (comparator = 'LIKE ANY')) ||
```
Comment (mickhansen) [inline lib/dialects/abstract/query-generator.js]: Not the most readable piece of code in the world but get's the job done i suppose, and not sure i can come up with anything more elegant right now.

Would it be too much hassle to ask you to test all the ANY paths? ($like, $iLike, $notLike, $notILike)

File: test/unit/sql/where.test.js [hunk 3]
```diff
@@ -423,6 +423,16 @@ suite(Support.getTestDialectTeaser('SQL'), function() {
             });
           });
         });
+
+        suite('$like', function() {
+          testsql('userId', {
+            $like: {
+              any: ['foo', 'bar', 'baz']
```
Comment (mickhansen) [inline test/unit/sql/where.test.js]: Please change the operator to `$any` both in code and tests.

File: grep [hunk 4]
```diff
@@ -0,0 +1 @@
...8471,"lng":-106.478627},{"lat":35.778974,"lng":-106.270068}],"color":"#a24ac3"},{"type":"polygon","latLngs":[{"lat":32.38471,"lng":-106.478627},{"lat":35.778974,"lng":-106.270068},{"lat":36.0384807732436,"lng":-105.38331627845764}],"color":"#a24ac3"},{"type":"polygon","latLngs":[{"lat":35.778974,"lng":-106.270068},{"lat":35.599283,"lng":-107.095915},{"lat":32.385761,"lng":-106.478705}],"color":"#a24ac3"},{"type":"polygon","latLngs":[{"lat":35.778974,"lng":-106.270068},{"lat":36.038477,"lng":-105.383302},{"lat":32.385863,"lng":-106.47839}],"color":"#a24ac3"},{"type":"polygon","latLngs":[{"lat":35.778711,"lng":-106.270406},{"lat":35.599283,"lng":-107.095915},{"lat":32.385761,"lng":-106.478705}],"color":"#a24ac3"},{"type":"polygon","latLngs":[{"lat":35.775864,"lng":-106.269219},{"lat":36.038192,"lng":-105.383517},{"lat":32.385863,"lng":-106.47839}],"color":"#a24ac3"},{"type":"polygon","latLngs":[{"lat":35.778974,"lng":-106.270068},{"lat":36.038477,"lng":-105.383302},{"lat":32.38471,"lng":-106.478627}],"color":"#a24ac3"},{"type":"polygon","latLngs":[{"lat":35.599283,"lng":-107.095915},{"lat":36.038477,"lng":-105.383302},{"lat":32.38471,"lng":-106.478627}],"color":"#a24ac3"}] -c polygon
```
Comment (mickhansen) [inline grep]: Doesn't look like this file should be here.

**Comment by mickhansen** [comment]
Could you squash this to a single descriptiev commit @3ch01c?

File: test/unit/sql/where.test.js [hunk 5]
```diff
@@ -423,6 +423,37 @@ suite(Support.getTestDialectTeaser('SQL'), function() {
...
+            }
+          }, {
+            postgres: "\"userId\" ILIKE ANY (ARRAY['foo','bar','baz'])"
+          });
+          testsql('userId', {
+            $notLike: {
+              $any: ['foo', 'bar', 'baz']
+            }
+          }, {
+            postgres: "\"userId\" NOT LIKE ANY (ARRAY['foo','bar','baz'])"
```
Comment (mickhansen) [inline test/unit/sql/where.test.js]: ARRAY does not need to be wrapped in paranthesis i assume?

**Comment by BridgeAR** [comment]
Ping @3ch01c

File: test/unit/sql/where.test.js (hunk 5 above)
Comment (3ch01c) [inline test/unit/sql/where.test.js]: I think the query generator adds them, but I don't think Postgres requires them.

File: lib/dialects/abstract/query-generator.js [hunk 6]
```diff
@@ -1766,6 +1767,11 @@ module.exports = (function() {
         value = (value.$between || value.$notBetween).map(function (item) {
           return self.escape(item);
         }).join(' AND ');
+      } else if (value && ((value.$like && (any = value.$like.$any) && (comparator = 'LIKE ANY')) ||
```
Comment (janmeier) [inline lib/dialects/abstract/query-generator.js]: This code is not very readable! It took me time to figure out that the if cases were actually also assignments

File: lib/dialects/abstract/query-generator.js (hunk 2 above)
Comment (janmeier) [inline lib/dialects/abstract/query-generator.js]: Perhaps here :https://github.com/3ch01c/sequelize/blob/likeany/lib/dialects/abstract/query-generator.js#L1780

```js
if _.isPlainObject(value) && value.$any 
  comparator += ' ANY';
  value = value.$any
```

**Comment by janmeier** [comment]
Thanks @3ch01c :)

**Comment by aesopwolf** [comment]
This doesn't seem to work for postgres, given ```$like: { $any: ['cat', 'hat']}``` then sequelize will generate 

```ILIKE ANY ARRAY['cat', 'hat']::VARCHAR```

but postgres wants the following:

```ILIKE ANY (ARRAY['cat', 'hat'])```

**Pull request merged**